Development

denishrana2 opened the pull request.

sahil030804 committed:
feat : GET All Books Info

Retrieve all books information from api by URL http://localhost:8000/api/books using GET method

sahil030804 committed:
feat : GET Single Book Information

Retrieve single book information by its book id which is send with params '/:id' in url like  http://localhost:8000/api/book/1  using GET method

sahil030804 committed:
Merge pull request #1 from sahil030804/feat/GetBooks

Feat : Get Books Information

sahil030804 committed:
feat : POST Book Data

Create/send new single book data from request body in json format and then pushes this new book data into books object with the help of POST method and then display newest added book data

sahil030804 committed:
Merge pull request #2 from sahil030804/feat/PostBook

feat : POST Book Data

sahil030804 committed:
feat : PUT/EDIT Book Data

Edit single book data by id which is send by url to fetch single book info which id is same as passes id and then change its name and price using PUT method also put validation for if name is available or not in body if available then change book name with body's book name and same for price vice versa

sahil030804 committed:
Merge pull request #3 from sahil030804/feat/PutBook

feat : PUT/EDIT Book Data

sahil030804 committed:
feat : Delete Book data by Id

create feature for Delete book data which id is match with url params id like first find index number of provided id in url params and then id index is available then delete it using splice method from bookindex to 1 element

sahil030804 committed:
Merge pull request #4 from sahil030804/feat/DeleteBook

feat : Delete Book data by Id

denishrana2 commented on package.json:
```diff
@@ -0,0 +1,16 @@
+  "main": "index.js",
+  "type": "module",
+  "scripts": {
+    "start": "nodemon --env-file=.env server.js"
```
Remove this env config line.

Use `dotenv` and `dotenv-safe`.
Add .env.example file too.

denishrana2 commented on server.js:
```diff
@@ -0,0 +1,22 @@
+import express from "express";
+import books from "./src/Routes/BookRoutes.js"; // import module from './src/Routes/BookRoutes.js'
+
+const PORT = process.env.PORT; // access port from env file
```
Add alter port (static) if no port is provided.

sahil030804 replied:
const PORT = process.env.PORT || 8080;

sahil030804 replied:
check in fix/bookApi branch

denishrana2 commented on server.js:
```diff
@@ -0,0 +1,22 @@
+app.get("/api/books", books);
+
+// Uses route module for access all routes after this url ('/api/book')
+app.use("/api/book", books);
```
Use only `/api` here.

books and book will be inside `bookRoutes.js` file. change accordingly.

sahil030804 replied:
check in fix/bookApi branch

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+];
+
+// FOR GET ALL BOOKS DATA
+router.get("/api/books", (req, res) => {
```
In `server.js` there is already `app.get("/api/books", books);` written.

What route is working in the end? and why to write this two places?

sahil030804 replied:
remove krta bhulay gyu pelama change krya p6i km k aa code me akho 2 thi 3 var lkhyo github na lidhe starting ma pull ne avu ky na kryu to bdhi file udi gyi branch create kri ama 😅😐

denishrana2 commented on server.js:
```diff
@@ -0,0 +1,22 @@
+import express from "express";
+import books from "./src/Routes/BookRoutes.js"; // import module from './src/Routes/BookRoutes.js'
```
the way you have written code in BookRoutes.js , it should be called bookService.js .

Because routes file are only used for routing. Controller and Service used afterwards.

sahil030804 replied:
ama idea na avyo su kryu ae me.
means k file nu name me wrong rkhyu 6 m?

denishrana2 replied:
ama tare Nodejs Folder Structure upar haji RND karvi padse.
Same Vasu ae pan ae j bhool kareli che.

See more youtube videos, read blogs on this.

sahil030804 replied:
haa because aa controller and validation ne ae bdhu alg files ma mukine use krvanu hve avyu video ma atle pela idea noto avyo

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+    return res.status(200).json(book);
+  }
+
+  res.status(404).json({ error: "User not found" });
```
kya thi copy karyu?

Error ma "User not found" che, means kasse thi uchkelu hoi avu lage che.

sahil030804 replied:
<img width="364" alt="Screenshot 2024-10-02 125527" src="https://github.com/user-attachments/assets/45e3c577-1285-4698-8f4b-57321de485ec">
acha are utaval ma user lkhay gyu pela user na j bnavya ta to 😅

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+
+// FOR GET SINGLE BOOK DATA BY ID
+router.get("/:id", (req, res) => {
+  let id = parseInt(req.params.id);
```
id string kari nakh. aapne dar vakhte parseInt karvani jaroor nai padse etle.

sahil030804 replied:
pn atyre ae string ma ley 6 to match ny thtu to '==' use kru to thy jse but datatype match ny thy ne to chalse?

sahil030804 replied:
okay i understand

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  // check price is not a number and if not give error if Book price must be digits
+  if (isNaN(Object.values(newBook)[2])) {
```
aa logic nai idea avyo.
Make it more readable. aa rite array values access nai karvani direct.
means, koi bija developer ne idea avo joiye ke kem lakhyu ae.

sahil030804 replied:
check in fix/bookApi branch

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+    if (name.trim() === "") {
+      return res.status(400).json({ error: "Book name can't be empty" });
+    } else {
+      book.name = name; // If valid, update the book name with body name
```
book update aa rite direct nathi karvanu.

use index based updating. means find index first, then je update karvu hoi ae array ma index target kari ne update karvanu.

sahil030804 replied:
check in fix/bookApi branch

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,112 @@
+
+  // Check if name is provided and not an empty string or whitespace
+  if (name) {
+    if (name.trim() === "") {
```
aa trim valu logic kone kidhu? jaate vicharyu? ke pachi kasse thi reference lidhelo che?

sahil030804 replied:
aa trim valu express JS no crash course valo video chalto to tyre hu jate practice krto tyre joyu tu atle yad htu to aa task ma implement kryu me validation mate.

sahil030804 committed:
fix/get all books code

fixes GET all books by fixes server.js file app.use('/api',books) and also change in bookroutes file to change endpoint to '/books'

sahil030804 committed:
Merge pull request #6 from sahil030804/fix/BookApi

fix/get all books code

sahil030804 committed:
fix : get single book data code

fix get single book data by change in bookroutes file with change endpoint '/book/:id'

sahil030804 committed:
Merge pull request #7 from sahil030804/fix/BookApi

fix : get single book data code

sahil030804 committed:
fix : get single book data by id code

fix get single book data by id but now it match string id with int id by '==' operator

sahil030804 committed:
Merge pull request #8 from sahil030804/fix/BookApi

fix : get single book data by id code

sahil030804 committed:
fix : post/create single book data code

fix post single book data by change in bookroutes file with change endpoint '/books'

sahil030804 committed:
Merge pull request #9 from sahil030804/fix/BookApi

fix : post/create single book data code

sahil030804 committed:
fix : change error of single book get

fix error of get single book data by changing 'Book not found'

sahil030804 committed:
Merge pull request #10 from sahil030804/fix/BookApi

fix : change error of single book get

sahil030804 committed:
fix : change update book data code

fix update book data by first it fetch bookindex of given id from params and then update data according to condition

sahil030804 committed:
Merge pull request #11 from sahil030804/fix/BookApi

fix : change update book data code

sahil030804 committed:
fix : change book price condition check in POST

fix POST method code of create book data. in that book price must be digit condition change from Object.values(newbook)[2] to newBook.price

sahil030804 committed:
Merge pull request #12 from sahil030804/fix/BookApi

fix : change book price condition check in POST

sahil030804 committed:
fix : change delete book data code

fix endpoint to /book/:id and also change id intostring again and change matching condition with '==' etc.

sahil030804 committed:
Merge pull request #13 from sahil030804/fix/BookApi

fix : change delete book data code

sahil030804 committed:
feat/fix : add static port

add static port number for access server if env file port in can't accesible

sahil030804 committed:
Merge pull request #14 from sahil030804/fix/BookApi

feat/fix : add static port

sahil030804 committed:
fix : add dotenv and dotenv-safe

add dotenv and dotenv-safe with .env.example file too.

sahil030804 committed:
fix : BookRoutes file name fix

Fixes BookRoutes.js file name with bookRoutes.js

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,140 @@
+  let newBook = {
+    id: books.length + 1,
+    name: name,
+    price: parseFloat(price),
```
jetli jagya ae parseFloat lakhyu che ae kadhi nakh. koi jaroor nai.

denishrana2 commented on src/Routes/BookRoutes.js:
```diff
@@ -0,0 +1,140 @@
+  let { name, price } = req.body; // create variables from req.body
+
+  let id = req.params.id;
+  let bookIndex = books.findIndex((book) => book.id == id);
```
Double equal valu nathi karvanu.

Te upar id: 1, id: 2 - avu lidhelu che. ama id: "1", id: "2" - avi rite lai le.
Etle problem solved. and use triple equals always.

sahil030804 committed:
fix : change folder structure of book api

fix book api project folder structure with also code and now routes are only in routes file , services/main logics are in service file and controllers are in controller file

sahil030804 committed:
Merge pull request #15 from sahil030804/fix/BookApi

Fix/book api

sahil030804 committed:
fix : change update api code and add little bit yup library

change update book api code with some validation if name and price is available then change according to body data otherwise keep is old same and also add yup library in code for validate book schema in post book api and create validation middleware

sahil030804 committed:
Merge pull request #16 from sahil030804/fix/BookApi

fix : change update api code and add little bit yup library

sahil030804 committed:
fix : change book api folder structure

change book api folder structure by create book folder which has all book related controller,services,routes etc. and also delete some files that is unnecessory right now and also there is little bit validation right now

sahil030804 committed:
Merge pull request #17 from sahil030804/fix/BookApi

fix : change book api folder structure

sahil030804 committed:
validation folder delete

sahil030804 committed:
Merge pull request #18 from sahil030804/fix/BookApi

validation folder delete

sahil030804 committed:
fix : add yup library for validations

add yup library for validate body data by the schema which is created in bookValidation file and create idNotFound middleware to check if id found or not then validate data and then run services file

sahil030804 committed:
fix : add JOI library for validation replace YUP

change yup library into joi library for server side validation so there is also some changes in validation schema and validation middleware for displaying custom error which we used to display error

sahil030804 committed:
Merge pull request #19 from sahil030804/fix/BookApi

Fix/book api

sahil030804 committed:
fix : remove id not found custom error msg from code

sahil030804 committed:
Merge pull request #20 from sahil030804/fix/BookApi

fix : remove id not found custom error msg from code
